### ISSUE TITLE
Chapter 9 linear regression using Gradient Descent mse does not change

### ISSUE THREAD

**Issue by larryfu** [ageron/handson-ml#124]
when run the linear regression Using Batch Gradient Descent
I got this result

  2017-11-17 21:49:10.939909: I tensorflow/core/platform/cpu_feature_guard.cc:137] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
  ('Epoch', 0, 'MSE =', 2.7544272)
('Epoch', 100, 'MSE =', 2.7544272)
('Epoch', 200, 'MSE =', 2.7544272)
('Epoch', 300, 'MSE =', 2.7544272)
('Epoch', 400, 'MSE =', 2.7544272)
('Epoch', 500, 'MSE =', 2.7544272)
('Epoch', 600, 'MSE =', 2.7544272)
('Epoch', 700, 'MSE =', 2.7544272)
('Epoch', 800, 'MSE =', 2.7544272)
('Epoch', 900, 'MSE =', 2.7544272)

I found mse and theta does not change at all 
system   ubuntu 16.04 
tensorflow   1.4.0
Is there some point I missed?

**Comment by ageron** [comment]
Hi @larryfu ,
Thanks for your question. I just verified (with TF 1.3 and also with TF 1.4), and it works fine for me. I really don't see anything wrong with the code, so I'm not sure what the problem is.  Could you please make sure you are using the latest notebook, without any changes, then restart the Jupyter kernel and run it from the start?

**Comment by larryfu** [comment]
@ageron Thank you very much for replying
I run it  with python directly

larry$ python linear_regression_gradients.py 

And here is the code 

import numpy as np
import tensorflow as tf
from sklearn.datasets import fetch_california_housing
from sklearn.preprocessing import StandardScaler

n_epochs = 1000
learning_rate = 0.01
housing = fetch_california_housing()
m, n = housing.data.shape
scaler = StandardScaler()
scaled_housing_data = scaler.fit_transform(housing.data)
scaled_housing_data_plus_bias = np.c_[np.ones((m, 1)), scaled_housing_data]

X = tf.constant(scaled_housing_data_plus_bias, dtype=tf.float32, name="X")
y = tf.constant(housing.target.reshape(-1, 1), dtype=tf.float32, name="y")
theta = tf.Variable(tf.random_uniform([n + 1, 1], -1.0, 1.0, seed=42), name="theta")
y_pred = tf.matmul(X, theta, name="predictions")
error = y_pred - y
mse = tf.reduce_mean(tf.square(error), name="mse")
gradients = 2/m * tf.matmul(tf.transpose(X), error)
training_op = tf.assign(theta, theta - learning_rate * gradients)
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    for epoch in range(n_epochs):
        sess.run(training_op)
        if epoch % 100 == 0:
            print("Epoch", epoch, "MSE =", mse.eval())

    best_theta = theta.eval()
    print  (best_theta)

**Comment by ageron** [comment]
I found the problem. I ran your code on Python 3 and it worked fine. Then I ran it on Python 2 and I got the same problem as you.  You are missing the following line at the beginning of your script. It's an import that is present at the beginning of the notebook:

```python
from __future__ import division, print_function, unicode_literals
````
The important part here is the division: it tells Python 2 to handle divisions just like in Python 3.  In Python 3, the `/` operator performs a float division (for example `9 / 2` is equal to 4.5), while the `//` operator performs an integer division (for example `9 // 2` is equal to 4). However, in Python 2, both operators perform an integer division, so `9 / 2` is equal to `9 // 2` which is equal to 4. Now consider this line in the script:

```python
gradients = 2/m * tf.matmul(tf.transpose(X), error)
````

When Python 2 computes `2/m`, it gets 0 as a result, so there's no gradients at all.  You could fix this by replacing this with `2.0/m`, but I personally prefer to add the `__future__` import, so I don't have to worry about this sort of thing when I'm coding.

Hope this helps!
Aurélien

**Comment by ruyingghost** [comment]
@larryfu 
I run the code above with TF1.3. The results are:

Epoch 0 MSE = 2.66595
Epoch 100 MSE = 0.630467
Epoch 200 MSE = 0.572585
Epoch 300 MSE = 0.558389
Epoch 400 MSE = 0.54899
Epoch 500 MSE = 0.542231
Epoch 600 MSE = 0.537337
Epoch 700 MSE = 0.533792
Epoch 800 MSE = 0.531221
Epoch 900 MSE = 0.529355
[[  2.06855226e+00]
 [  7.74078071e-01]
 [  1.31192386e-01]
 [ -1.17845096e-01]
 [  1.64778158e-01]
 [  7.44080753e-04]
 [ -3.91945168e-02]
 [ -8.61356616e-01]
 [ -8.23479712e-01]]

Process finished with exit code 0

**Comment by ageron** [comment]
@ruyingghost , thanks for confirming that the code works well. I'm guessing you are using Python 3. The problem was due to a missing import (`from __future__ import division`), see my answer above.

**Comment by larryfu** [comment]
It  works!  thank you @ageron  and  @ruyingghost 
I'm new to Machine learning and python, It helps me a lot .
